cli_tools/common/utils/daisy/resource_labeler.go

```diff
@@ -0,0 +1,98 @@
+//  Copyright 2019 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package daisyutils
+
+import "github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+
+// ResourceLabeler is responsible for labelling GCE resources (instances, disks and images) with
+// labels used to track resource creation by import processes.
+type ResourceLabeler struct {
+	BuildID                   string
+	UserLabels                *map[string]string
+	BuildIDLabelKey           string
+	InstanceLabelKeyRetriever InstanceLabelKeyRetrieverFunc
+	DiskLabelKeyRetriever     DiskLabelKeyRetrieverFunc
+	ImageLabelKeyRetriever    ImageLabelKeyRetrieverFunc
+}
+
+// InstanceLabelKeyRetrieverFunc returns GCE label key to be added to given instance
+type InstanceLabelKeyRetrieverFunc func(image *daisy.Instance) string
+
+// DiskLabelKeyRetrieverFunc returns GCE label key to be added to given disk
+type DiskLabelKeyRetrieverFunc func(image *daisy.Disk) string
+
+// ImageLabelKeyRetrieverFunc returns GCE label key to be added to given image
+type ImageLabelKeyRetrieverFunc func(image *daisy.Image) string
+
+// LabelResources labels workflow resources temporary and permanent resources with appropriate
+// labels
+func (rl *ResourceLabeler) LabelResources(workflow *daisy.Workflow) {
+	for _, step := range workflow.Steps {
```

Comment:
I wonder whether there is a way to lock these resources before workflow is ended. Label is a hint but not a enforced rule to prevent accident deletion.

Reply:
This is something we can have a look if we decide not to implement a service. With a service, all this will be hidden from users since we will run imports outside user projects. Feel free to add a tracking bug.